File: leakcanary-analyzer/src/main/java/com/squareup/leakcanary/LeakTrace.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.leakcanary;

import android.support.annotation.NonNull;
import java.io.Serializable;
import java.util.List;

/**
 * A chain of references that constitute the shortest strong reference path from a leaking instance
 * to the GC roots. Fixing the leak usually means breaking one of the references in that chain.
 */
public final class LeakTrace implements Serializable {

  @NonNull public final List<LeakTraceElement> elements;
  @NonNull public final List<Reachability> expectedReachability;

  LeakTrace(List<LeakTraceElement> elements, List<Reachability> expectedReachability) {
    this.elements = elements;
    this.expectedReachability = expectedReachability;
  }

  @Override public String toString() {
    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < elements.size(); i++) {
      LeakTraceElement element = elements.get(i);
      sb.append("* ");
      if (i != 0) {
        sb.append("↳ ");
      }
      boolean maybeLeakCause = false;
      Reachability currentReachability = expectedReachability.get(i);
      if (currentReachability == Reachability.UNKNOWN) {
        maybeLeakCause = true;
      } else if (currentReachability == Reachability.REACHABLE) {
        if (i < elements.size() - 1) {
          Reachability nextReachability = expectedReachability.get(i + 1);
          if (nextReachability != Reachability.REACHABLE) {
            maybeLeakCause = true;
          }
        } else {
          maybeLeakCause = true;
        }
      }
      sb.append(element.toString(maybeLeakCause)).append("\n");
    }
    return sb.toString();
  }

  public @NonNull String toDetailedString() {
    String string = "";
    for (LeakTraceElement element : elements) {
      string += element.toDetailedString();
    }
    return string;
  }
}

File: leakcanary-analyzer/src/main/java/com/squareup/leakcanary/LeakTrace.kt
package com.squareup.leakcanary

import com.squareup.leakcanary.Reachability.Status.REACHABLE
import com.squareup.leakcanary.Reachability.Status.UNKNOWN
import com.squareup.leakcanary.Reachability.Status.UNREACHABLE
import java.io.Serializable

/**
 * A chain of references that constitute the shortest strong reference path from a leaking instance
 * to the GC roots. Fixing the leak usually means breaking one of the references in that chain.
 */
data class LeakTrace(
  val elements: List<LeakTraceElement>,
  val expectedReachability: List<Reachability>
) : Serializable {

  override fun toString(): String {
    var leakInfo = "┬" + "\n"
    val lastElement = elements.last()
    val lastReachability = expectedReachability.last()
    elements.dropLast(1)
        .forEachIndexed { index, leakTraceElement ->
          val currentReachability = expectedReachability[index]

          leakInfo += """
        |├─ ${leakTraceElement.className}
        |│${getReachabilityString(currentReachability)}
        |│${getPossibleLeakString(currentReachability, leakTraceElement, index)}
        |
        """.trimMargin()
        }
    leakInfo += """╰→ ${lastElement.className}
      |$ZERO_WIDTH_SPACE${getReachabilityString(lastReachability)}
    """.trimMargin()

    return leakInfo
  }

  fun toDetailedString(): String {
    return elements.joinToString {
      it.toDetailedString()
    }
  }

  private fun getPossibleLeakString(
    reachability: Reachability,
    leakTraceElement: LeakTraceElement,
    index: Int
  ): String {
    val maybeLeakCause = when (reachability.status) {
      UNKNOWN -> true
      REACHABLE -> {
        if (index < elements.lastIndex) {
          val nextReachability = expectedReachability[index + 1]
          nextReachability.status != REACHABLE
        } else {
          true
        }
      }
      else -> false
    }
    return DEFAULT_NEWLINE_SPACE + "↓" + " ${leakTraceElement.toString(maybeLeakCause)}"
  }

  private fun getReachabilityString(reachability: Reachability): String {
    return DEFAULT_NEWLINE_SPACE + "Leaking: " + when (reachability.status) {
      UNKNOWN -> "UNKNOWN"
      REACHABLE -> "NO (${reachability.reason})"
      UNREACHABLE -> "YES (${reachability.reason})"
    }
  }

  companion object {
    private val DEFAULT_NEWLINE_SPACE = "                 "
    private val ZERO_WIDTH_SPACE = '\u200b'
  }
}